clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -5024,11 +5023,25 @@ bool SYCLIntegrationFooter::emit(raw_ostream &OS) {
   Policy.SuppressTypedefs = true;
   Policy.SuppressUnwrittenScope = true;
 
+  llvm::SmallVector<const VarDecl *> VisitedSpecConstants;
+
   // Used to uniquely name the 'shim's as we generate the names in each
   // anonymous namespace.
   unsigned ShimCounter = 0;
   for (const VarDecl *VD : SpecConstants) {
     VD = VD->getCanonicalDecl();
+
+    // Skip if we've already visited this.
+    if (llvm::find(VisitedSpecConstants, VD) != VisitedSpecConstants.end())
+      continue;
+
+    // Skip if this isn't a SpecIdType.  This can happen if it was a deduced
+    // type.
+    if (!Util::isSyclSpecIdType(VD->getType().getCanonicalType()))
+      continue;
```

Comment:
Do we expect to see a lot of non-specialization_id types here? If so, I suggest to swap those two checks to avoid extra search in the `vector` for things which won't appear there ever.